sdk/helper/keysutil/policy.go

```diff
@@ -1973,3 +1951,160 @@ func (p *Policy) EncryptWithFactory(ver int, context []byte, nonce []byte, value
 
 	return encoded, nil
 }
+
+func (p *Policy) UpdateKeyVersion(ctx context.Context, storage logical.Storage, key []byte, isPrivateKey bool, keyVersion int) error {
+	keyEntry, err := p.safeGetKeyEntry(keyVersion)
+	if err != nil {
+		return err
+	}
+
+	// Validations
+	if !p.Type.ImportPublicKeySupported() {
+		return errors.New("provided type does not support version updates")
+	}
+
+	publicKeyImported := keyEntry.PublicKeyImported
+	if publicKeyImported && !isPrivateKey {
+		// NOTE: Description
+		return errors.New("cannot import a public key to a key version that already as a public key set")
+	}
+
+	if !publicKeyImported {
+		// NOTE: Description
+		return errors.New("private key imported, cannot be updated or add a public key")
+	}
+
+	// Parse key
+	parsedPrivateKey, err := x509.ParsePKCS8PrivateKey(key)
+	if err != nil {
+		return fmt.Errorf("error parsing asymmetric key: %s", err)
+	}
+
+	// Check if key pair is valid, this reduces the need for duplicated code but moves the checks to after validating key pair
+	switch parsedPrivateKey.(type) {
+	case *ecdsa.PrivateKey:
+		ecdsaKey := parsedPrivateKey.(*ecdsa.PrivateKey)
+		// Is this going to work?
+		publicKey, err := x509.ParsePKIXPublicKey([]byte(keyEntry.FormattedPublicKey))
+		if err != nil {
+			return fmt.Errorf("failed to parse key entry public key: %v", err)
+		}
+		if !publicKey.(*ecdsa.PublicKey).Equal(ecdsaKey.PublicKey) {
+			// NOTE: Description
+			return fmt.Errorf("keys do not match")
+		}
+	case *rsa.PrivateKey:
+		rsaKey := parsedPrivateKey.(*rsa.PrivateKey)
+		if !rsaKey.PublicKey.Equal(keyEntry.RSAPublicKey) {
+			// NOTE: Description
+			return fmt.Errorf("keys do not match")
+		}
+	}
+
+	err = p.parseKeyToKeyEntry(&keyEntry, parsedPrivateKey, isPrivateKey)
+	if err != nil {
+		return err
+	}
+
+	p.Keys[strconv.Itoa(keyVersion)] = keyEntry
+
+	return p.Persist(ctx, storage)
+}
+
+// NOTE: func name?/not a fan of changing input params
```

Comment:
Ah, its the C programmer in me, should've seen some of my initial PKI related code ;-) 

If you want, perhaps it would be cleaner to write this as `func (e *keyEntry) parseFromKey(...)`? Then you don't have the return semantics to deal with, and you can pass in what little you use from `p` into the alternative equivalent thingy.

Just a thought :-)